### ISSUE TITLE
$ doesn't match CRLF

### ISSUE THREAD

**Issue by jminer** [rust-lang/regex#244]
I created a regex with `multi_line` set to `true`, but after debugging why the regex was matching in a unittest but not in a file, I found out that `$` isn't matching the end of a line in the file. I'm using Windows so the newlines are `\r\n`.

**Comment by BurntSushi** [comment]
Could you please provide a test case that doesn't act as you expect?

**Comment by jminer** [comment]
Sure, here is a program that I would expect to print "Matched: true", but it prints "Matched: false":

    extern crate regex;
    
    use regex::RegexBuilder;
    
    fn main() {
        let regex = RegexBuilder::new("^apple$").multi_line(true).compile().unwrap();
        let text = "\r\napple\r\nbanana";
        let mut matched = false;
        for _ in regex.captures_iter(text) {
            matched = true;
        }
        println!("Matched: {}", matched);
    }

**Comment by BurntSushi** [comment]
That is indeed expected behavior. `$` will only match `\n` in multi line mode. It's not clear to me whether supporting `\r\n` is feasible unfortunately.

**Comment by BatmanAoD** [comment]
@BurntSushi Would treating `\r` as an end-of-line character, and `\n` as *non*-EOL if preceded by `\r`, be an acceptable change? This should be doable in a DFA, I think, though it does mean two EOL states. It might be somewhat surprising behavior when `\r` is embedded in a line, but that seems like a much rarer case than EOL `\r`, and it's not even clear to me that treating carriage return as EOL is actually wrong.

**Comment by BurntSushi** [comment]
@BatmanAoD That sounds feasible from an implementation perspective, but I'm not a fan of implementing something that's wrong. (Essentially no systems use `\r` for line endings any more, and on Windows, it's `\r\n`, not `\r`.)

**Comment by BatmanAoD** [comment]
But the existing implementation is *more* wrong, so I'm not sure I understand that as an objection.

**Comment by BatmanAoD** [comment]
(Also, just last week I actually did run into something that uses `\r` on its own as EOL by default: Putty in serial mode! I was shocked.)

**Comment by mitsuhiko** [comment]
I just ran into this when trying to use it to parse some data coming in via HTTP. This is incredibly confusing at the very leasts as this is the only regex engine I know with this behavior.

**Comment by BurntSushi** [comment]
I believe RE2 also has this behavior. `\r?$` may be a not-ideal work around.

**Comment by BurntSushi** [comment]
I'm going to re-open this, but I don't have any immediate plans to work on it.

**Comment by mitsuhiko** [comment]
Just wanted to throw in that I was wrong. It's indeed the same behavior in Python and Go as well. I did never notice in the former because a similar code I was working with stripped the EOL chars whereas the API I used in rust only split on \n and left an empty \r hanging around.

So when parts of the data were recombined into a string later the \r were left in there.

**Comment by mitsuhiko** [comment]
So the regex behavior here does match other engines dispite what I said earlier.

**Comment by BurntSushi** [comment]
@mitsuhiko Oh interesting. I should have known for Go, but it's interesting to see that Python doesn't do it either:

```
>>> import re
>>> re.match('foo$', 'foo\n') is not None
True
>>> re.match('foo$', 'foo\r\n') is not None
False
```

So I guess we're in good company?

**Comment by BatmanAoD** [comment]
@BurntSushi That's odd. I've used `re` in Python 2.7 on Windows pretty extensively, and I'm sure I've used `$`, and I know the files I was working with had carriage returns in them. I would have thought I'd notice this! I suppose I must have stripped all my input lines before searching for patterns.

**Comment by BurntSushi** [comment]
@BatmanAoD In Python, I believe if you open your files with `U` (universal line mode?), then Python will do something clever automatically. Splitting by line and then searching will probably also do it.

**Comment by BurntSushi** [comment]
`/foo$/m` in Javascript does match `foo\n` and `foo\r\n`.

**Comment by jminer** [comment]
Java's regex also matches `\r\n` by default:

    Pattern p = Pattern.compile("foo$", Pattern.MULTILINE);
    System.out.println(p.matcher("foo\r\n").find());

prints `true`.

**Comment by BatmanAoD** [comment]
For Javascript and Java, does `\r` on its own match as a newline anchor?

`\r` on its own is apparently considered a newline character by Unicode.

Is there any objection to simply treating [every Unicode line terminator character sequence](https://en.wikipedia.org/wiki/Newline#Unicode) as a match for `$`? @BurntSushi, I know you said earlier that treating `\r` on its own as a line terminator would be "wrong", but I'm still not quite sure I see why you'd consider that to be incorrect behavior, even if it's not the norm for regex engines.

**Comment by mitsuhiko** [comment]
JavaScript matches them alone:

```javascript
> '1\n2\r3\r\n4'.split(/$/m)
[ '1', '\n2', '\r3', '\r', '\n4' ]
```

**Comment by BatmanAoD** [comment]
@mitsuhiko Hmm. If the interpreter [here](http://eloquentjavascript.net/code/) is correct, JavaScript also returns a four-element array here: `'1\r\n\r\n4'.split(/$/m)`, which is obviously not correct on Windows (there are only two line-endings there).

**Comment by mitsuhiko** [comment]
@BatmanAoD which browser are you using? Chrome, Firefox and Safari gives me 5 elements. Same with JavaScript Core and V8 in node.

**Comment by BatmanAoD** [comment]
Sorry, I meant 5 elements, splitting on each of the `\r`'s and `\n`'s (so 4 matches, which is what I was thinking when I typed that).

There should only be 3 elements (2 matches).

**Comment by mitsuhiko** [comment]
I'm not sure. The above behavior makes perfect sense if you go by unicodes classification of newline characters. I find that behavior quite good because it means that matching with `$` works in all newline environments. I know some people say files ending with `\r` are not common any more but if anyone ever worked with OS X knows that `\r` newlines are a thing of the present. I come across such files regularly.

**Comment by BatmanAoD** [comment]
I don't believe you're interpreting that list of Unicode newline character-sequences correctly, because they list `\r\n` as a *separate entry* from `\r`. I.e., Unicode considers `\r\n` to be *one single* newline (as they must, since Windows is so widely used).

A common pattern is to use `^$` to find blank lines; this would give 100% false positives using the JavaScript behavior.

**Comment by BatmanAoD** [comment]
Java does exactly the right thing:

    Pattern p = Pattern.compile("^$", Pattern.MULTILINE);
    Matcher m = p.matcher("1\r\r\n2\r\n\r\n3");
    while (m.find()) { System.out.println("match at " + m.start() + ": " + m.group()); } 

This prints:

    match at 2: 
    match at 7: 

I.e., the first `\r` is treated as a newline, after which each group of `\r\n` together is a single newline.

**Comment by BatmanAoD** [comment]
(To be more precise: Java's behavior is, as far as I can tell, precisely equivalent to the behavior we'd get from implementing my proposal of treating `\r` like a newline all the time and `\n` like a newline *except* when preceded by `\r`.)

**Comment by mitsuhiko** [comment]
Unicode does not define control characters. Unicode only has recommendations on newline handling and the recommendations and those would be "convert from platform newline characters to LS or PS" and then back which I think nobody does. So I think unicode in itself is unclear on it. However unicode has guidelines on regular expressions:

These two things apply:

> Line Boundaries
To meet this requirement, if an implementation provides for line-boundary testing, it shall recognize not only CRLF, LF, CR, but also NEL (U+0085), PARAGRAPH SEPARATOR (U+2029) and LINE SEPARATOR (U+2028).

as well as

> It is strongly recommended that there be a regular expression meta-character, such as "\R", for matching all line ending characters and sequences listed above (for example, in #1). This would correspond to something equivalent to the following expression. That expression is slightly complicated by the need to avoid backup.
>
> `(?:\u{D A}|(?!\u{D A})[\u{A}-\u{D}\u{85}\u{2028}\u{2029}]`
>
> Note: For some implementations, there may be a performance impact in recognizing CRLF as a single entity, such as with an arbitrary pattern character ("."). To account for that, an implementation may also satisfy R1.6 if there is a mechanism available for converting the sequence CRLF to a single line boundary character before regex processing.

WRT to Java behavior from above it yields this (in pseudocode):

```javascript
> '1\n2\r3\r\n4'.split(/$/m)
['1', '\n2', '\r3', '\r\n4']
```

**Comment by mitsuhiko** [comment]
Also something to add to this CRLF business:

> Arbitrary character pattern (often ".")
Where the 'arbitrary character pattern' matches a newline sequence, it must match all of the newline sequences, and \u{D A} (CRLF) should match as if it were a single character. (The recommendation that CRLF match as a single character is, however, not required for conformance to RL1.6.)
Note that ^$ (an empty line pattern) should not match the empty string within the sequence \u{D A}, but should match the empty string within the reversed sequence \u{A D}.

All of this is from here: http://unicode.org/reports/tr18/

I guess my recommendation would be to change the behavior to handle `\r` and `\n` as newline characters on their own right, to leave CRLF essentially unhandled and recommend people to normalize CRLF to LF before processing if they need to match them as a single character.

**Comment by mitsuhiko** [comment]
Also one more thing and I shut up: it turns out that .NET which is probably one of the highest authorities on windows newline handling has the same behavior as python, this crate and go: https://msdn.microsoft.com/en-us/library/yd1hzczs.aspx#Multiline

> By default, $ matches only the end of the input string. If you specify the RegexOptions.Multiline option, it matches either the newline character (\n) or the end of the input string. It does not, however, match the carriage return/line feed character combination. To successfully match them, use the subexpression \r?$ instead of just $.

**Comment by BatmanAoD** [comment]
I'm not sure what you mean by "Unicode does not define control characters", since it certainly specifies control character definitions (inheriting some from ASCII and introducing some others, such as the text-direction characters).

By treating `\r` as a newline but *not* matching `\n` as a newline when it immediately follows `\r`, it looks like we'd meet all the Unicode recommendations except causing `.` to match `\r\n` (which would be in a completely separate part of the input logic anyway, since `.` isn't a zero-width match).

Again, your recommendation of letting `\r\n` match *two separate* newlines (i.e. not including special handling for `\n` following `\r`) would have the wrong behavior for the `^$` pattern, which is explicitly called out in the text you quoted.

.......but the fact that Microsoft can't be bothered to get this right in .NET just blows me away. Ugh.

**Comment by mitsuhiko** [comment]
> I'm not sure what you mean by "Unicode does not define control characters"

That was in reference to the unicode guidelines on newline handling:

> The following table provides hexadecimal values for the acronyms used in the text […] The Unicode Standard does not formally assign control characters, instead it provides the 65 code values for use as in the 7 and 8-bit standards. See The Unicode Standard, Version 2.0, Section 2.6 Controls and Control Sequences.

For what it's worth I do not object to your recommendation to match `\r\n` together. Just quoting form the unicode guide on recommended practices. I am at this point probably happy with any behavior because it seems different engines have different ideas about this.

The only reason I ran into this is actually not this library but Rust's `BufRead::read_line` which does not support either `\r` or `\r\n`. So my original confusing was caused by a misunderstanding of the entire situation in the first place. So maybe it would be worth updating the Rust and/or regex docs to call out on this behavior.

**Comment by BurntSushi** [comment]
> I know you said earlier that treating \r on its own as a line terminator would be "wrong", but I'm still not quite sure I see why you'd consider that to be incorrect behavior, even if it's not the norm for regex engines.

It's wrong for the reasons you yourself pointed out? When you do a split, you wind up treating each `\r` and `\n` as a line terminator. I think you later added this criterion:

> By treating `\r` as a newline but not matching `\n` as a newline when it immediately follows `\r`

But this is distinctly harder than "just making `$` recognize `\r` as a line terminator."

I do find it very interesting that not even .NET gets this right. I think they use a backtracking engine, and it *should* be easier for a backtracking engine to support this feature. Maybe they goofed initially and now there's a backcompat concern.

**Comment by BatmanAoD** [comment]
> I think you later added this criterion...

No, no, I really did think through my proposal, I promise! :laughing: My original comment was:

> treating `\r` as an end-of-line character, and `\n` as *non*-EOL if preceded by `\r`...

Anyway, as for look-behind: this is only single-character lookbehind, which is already necessary for handling `\b` (which is why, in `input.rs`, the `Input` trait has a `previous_char` method). Though I grant you that correctly handling line-endings will almost certainly make searching _slower_, since `previous_char` has some cost (the implementation appears to re-decode the appropriate bytes every time the function is called) and the actual check for an `EndLine` match will be more complicated.

**Comment by BatmanAoD** [comment]
...anyway, the change sounds trivial to implement, but on my local machine I couldn't get `ripgrep` to pass the test case I added in my fork, so I'm probably doing something wrong. (This may be due to the fact that I love Rust in theory but haven't written more than about 30 lines of code in it so far.)

**Comment by BurntSushi** [comment]
> (which is why, in input.rs, the Input trait has a previous_char method)

The `regex` crate is a complex beast. The `Input` trait is only half the battle. It isn't used in the DFA, which has to encode the matching semantics into the DFA itself. In fact, the DFA cannot even handle the full Unicode variant of `\b`. (If `\b` is used and a non-ASCII byte is seen, then the DFA immediately quits and hands over control to one of the NFA engines.)

> the change sounds trivial to implement

It's not. I promise.

**Comment by BatmanAoD** [comment]
Sorry, looking at `input.rs` I figured it was probably used throughout as a unified interface to the search data--I didn't realize that the DFA engine doesn't actually use it. I also didn't realize there are multiple engines used.

So, if I understand correctly, to implement my proposal, `input.rs` would need to change, *and* all the engines used would need to change; based on the `HACKING` document, these engines are:

 * The primary engine (DFA)
 * The PikeVM
 * The Boyer-Moore engine
 * The Aho-Corasick engine (which is in a different crate)
 * The PikeVM implementation generated by `regex!`

...Is that correct?

`HACKING` also states:

> All of the above matching engines expose precisely the same matching semantics.
This is indeed tested. (See the section below about testing.)

....do we know for certain that all of the listed engines have the same behavior for `/r/n` and `$`?

**Comment by BatmanAoD** [comment]
...actually, one thing doesn't seem to be answered by the `HACKING` document--what *is* `input.rs` used for? Any engines that use `input.rs` would be automatically "fixed" by the proposed change to `input.rs`--right?

**Comment by BurntSushi** [comment]
@BatmanAoD The literal engines don't care about `^` and `$` (yet), so you don't need to worry about those. So that includes the "boyer-moore engine" (I'll have to update the HACKING guide because that name is wrong) and the Aho-Corasick engine can be ignored. (There are several other literal search strategies as well.) Your list is missing the backtracking engine.

So to be clear, in order to fix this, you need to:

* Fix handling of the `$` anchor in `input.rs`. Hopefully, this fixes the PikeVM and backtracking engines.
* Fix handling of the `$` anchor in `dfa.rs`. This fixes the DFA engine. (This is the hard bit.)

> what is input.rs used for? 

It's a strange abstraction for permitting the same NFA engines (PikeVM and backtracking) to run on both Unicode-aware and Unicode-unaware regexes.

**Comment by BatmanAoD** [comment]
Playing with the code a little, I believe the "reverse" searching mechanism may be a fairly significant hurdle to fixing this; could you please provide a little more info about what exactly this logic is for and how it works, or at least a pointer to where I should start looking to understand it?

From my earlier comments on BurntSushi/ripgrep#416:

> Is it just what's described in rust-lang/regex#190? Does the engine construct a DFA designed (guaranteed?) to find the same matches when applied backwards that the "normal" DFA would find when applied forwards?

> If so, then with regard to this particular issue, it might actually make my proposed fix infeasible (or at least more complicated than I'd like), unless I'm misunderstanding something. When running the DFA forward, `$` should anchor immediately before the current character-index if the character is `\n` and the previous character *isn't* `\r`, but in reverse, `$` should anchor after `\n` unless the *next* character to be observed (i.e. the one right before `\n` in the input file) isn't `\r`. Based on the precedent set by the word-boundary logic, the way to handle this multi-character anchor dependency is via a special flag in `StateFlags`. But the actual logic to assign the flags for the typical case is in `exec_byte`, which (as far as I can tell) is intentionally designed not to have any special behavior for the reverse match. Moreover, there's not really any way that I can see to look *ahead* in the search, which I think would be necessary in the reverse case.

**Comment by BurntSushi** [comment]
> could you please provide a little more info about what exactly this logic is for

It's for finding the start of a match using the DFA engine.

> how it works

The reverse DFA is just like the forward DFA, save two primary differences:

1. During compilation, a reverse DFA is emitted by following the same procedure as the forward DFA, except all concatenations are reversed. (There are more details to this, but reversing the concatenations is the big one.)
2. During execution, a reverse DFA is run from the end of a *guaranteed* match until a match is found. The position of this match corresponds to the starting position of a match.

The forward DFA matches *if and only if* the reverse DFA matches.

> or at least a pointer to where I should start looking to understand it?

Unfortunately, I don't know. I remember doing an exercise like this many moons ago in my computational theory class. Google returns various results but most of them are theory based. I don't know of any regex engine that does this other than RE2 and this one (my guess is that Hyperscan does as well, actually), so you might consider consulting RE2 for a slightly different perspective if that helps.

With respect to your comments... My deepest apologies, but I just can't usefully respond to them without loading a bunch of stuff about the DFA implementation into my brain's cache, and that will take at least a few hours that I can't afford to spend right now. (That I even have to do this is a failing of the implementation: it is too complex.)

**Comment by BatmanAoD** [comment]
> My deepest apologies, but I just can't usefully respond to them without loading a bunch of stuff about the DFA implementation into my brain's cache, and that will take at least a few hours that I can't afford to spend right now.

I definitely understand! I am also time-constrained, as you may be able to tell from my very sporadic engagement in these issue threads. :disappointed: 

 > My *intuition* is that if the reverse DFA is hard, then the forward DFA is also hard and there might be some cases you just aren't considering.

Yes, `\r\n` makes correctly matching `^` tricky in the forwards case, for the same reason (assuming that we want it to treat standalone `\r` as a newline, matching between `\r` and anything other than `\n`). I think I specifically found the reverse case trickier, though, for some implementation-specific reason--I'll have to look again when I next open up the project.

Thanks for your help!

**Comment by BatmanAoD** [comment]
I haven't had any new thoughts about how special handling for `\r\n` could be implemented, but is fixing this even a possibility now that `regex` has hit 1.0, or would it present too much of a backwards-compatibility risk?

(I doubt anyone relies on `$` matching the *middle* of the `\r\n` sequence, but still....)

**Comment by BurntSushi** [comment]
@BatmanAoD I thought about that before 1.0, but the conclusion is inescapable:

1. Blocking regex 1.0 on this issue doesn't make sense.
2. Release regex 2.0 _just_ for this issue doesn't seem worth it.
3. Introducing a new opt-in feature and/or flag is a backwards compatible change, and has precedent in other regex engines such as PCRE2 with respect to how line terminators are treated.

**Comment by phaazon** [comment]
Hello everybody,

Is there a workaround / fix for this issue? I’m hesitating on doing the “clean CRLF by hand before using `regex`”, but that seems like completely insane.

**Comment by BurntSushi** [comment]
This is what I do for ripgrep: https://docs.rs/grep-regex/0.1.3/grep_regex/struct.RegexMatcherBuilder.html#method.crlf

> “clean CRLF by hand before using regex”, but that seems like completely insane.

Not to me. Sounds like the best path for some scenarios.

**Comment by phaazon** [comment]
@BurntSushi so I guess this is not exposed currently in the `RegexBuilder` type on the latest version?

**Comment by BurntSushi** [comment]
@phaazon Sorry, but I don't know what you're talking about. I linked you to a work-around _outside_of the `regex` crate that makes sense in specific contexts. It is not part of the regex crate itself.

**Comment by phaazon** [comment]
Yeah, but question was more like: you have no plan to support that in the `regex` crate?

**Comment by BurntSushi** [comment]
@phaazon You asked for a _work-around_, and I told you what I know and things I've used.

If you're instead asking whether I plan to support this properly in the regex crate---which is an entirely different question---then yes, it would be nice. That's why this ticket is open. But it's non-trivial. See discussion above.

**Comment by BatmanAoD** [comment]
@phaazon 

> I’m hesitating on doing the “clean CRLF by hand before using regex”, but that seems like completely insane.

Note that this is actually what GNU grep does.

**Comment by phaazon** [comment]
![](https://giphy.com/embed/KHDIhwKO7LBG8)

**Comment by phaazon** [comment]
![](https://phaazon.net/media/uploads/shrug2.gif)

**Comment by BurntSushi** [comment]
@phaazon Please don't post meme images on any issue tracker that I maintain.

**Comment by jzabroski** [comment]
> > “clean CRLF by hand before using regex”, but that seems like completely insane.
> 
> Not to me. Sounds like the best path for some scenarios.

I mean, I guess if it's just for this one sigil, you have a great point.  But you can take this "best path" too far, too.  This is what IBM did for java.util.Regex: they wrote a Java pre-compiler that replaces ICU Regex with java.util.Regex, and... it's IBM.  

I guess if you're not going to fix this issue, at least update the documentation referencing TR18 to clarify that you don't handle "Unicode New Line" in R1.6 (which is what we've been discussing for .NET).

**Comment by BurntSushi** [comment]
@jzabroski Please consider reading my comment above: https://github.com/rust-lang/regex/issues/244#issuecomment-491930753 

And this crate's Unicode support _is_ pretty thoroughly documented, including this omission: https://github.com/rust-lang/regex/blob/master/UNICODE.md

> But you can take this "best path" too far, too.

Pretty much anything can be taken too far. Please don't misconstrue my comments on a particular issue as generic advice for any situation.

**Comment by jzabroski** [comment]
Yeah, I also found your Hacker News comments helpful: https://news.ycombinator.com/item?id=17083461

Thanks for pointing me to that link, again.  Sigh.  I read it via https://docs.rs/crate/regex/1.1.0/source/UNICODE.md but somehow my brain skipped over the fact you don't handle Unicode line boundaries.  I was rapidly trying to build a table of regex engines  and conformance to TR18 (via google searching "TR18 Regular Expression" etc.), so that might be why I missed this rather obvious call out.

**Comment by jzabroski** [comment]
By the way, sorry if my comment left you confused.

Do you happen to know which Unicode Regex libraries _do_ support Unicode New Line?  I went through icu4j source code, your source code, and a bunch of other places, and I don't know where Mark Davis came up with his guidance for what a Unicode New Line is, since nobody seems all that interested in portable New Line characters.

**Comment by BurntSushi** [comment]
@jzabroski I believe PCRE2 supports them. From `man pcre2api`:

> PCRE2  supports five different conventions for indicating line breaks in strings: a single CR (carriage return) character, a single LF (linefeed) character, the two-character sequence CRLF, any of the three preceding, or any Unicode newline sequence. The Unicode newline sequences are the three just mentioned, plus the single characters VT (vertical tab, U+000B), FF (form feed, U+000C), NEL (next line, U+0085), LS (line separator, U+2028), and PS (paragraph separator, U+2029).

It doesn't explicitly mention UTS#18, but that sounds like correct RL1.6 support to me.

Certainly, folks want CRLF support, but I don't think I've ever heard of anyone caring about more than that. Interestingly, for this regex engine, adding the rest of the Unicode line separators is easier than supporting `\r\n` since only the latter consists of two codepoints.

**Comment by jzabroski** [comment]
Thanks.  One last question (and you can email me if you want to take it offline).  How does PCRE2 do those conventions?  When I began authoring https://github.com/dotnet/corefx/issues/28410 I had thought that it might be nice to have `$` just be a "programmable sigil", but also felt that wasn't highly usable and would introduce very non-local code to the Regex reader.  Just wondering if PCRE2 had a non-Perl-like readable approach ;D

Anyway: Not familiar and don't know where to look for the PCRE2 code.  Your code, ICU, Python 3.7 regex module, and RogueWave's C++ code came up tops in Google Search.  I must confess, I didn't even know there was a PCRE _2.0_ until today.  Insert quote from _The Lathe of Heaven_ about dreams changing our reality.

**Comment by BurntSushi** [comment]
I confess that I do not know. I'm not familiar with PCRE2's internals, and when I have looked, it was pretty hard to understand anything. But I didn't try too hard.

With that said, I don't imagine that there is anything weird going on here? PCRE2 supports look-around already, and since RL1.6 can be implemented in terms of look-around, it seems straight-forward to me. Maybe I'm misunderstanding your question though?

**Comment by jzabroski** [comment]
Maybe.  My thought is, if several people are already regex replacing `$` with a fancier pattern to form their real regex, why not just have code that looks like this (C#):

```c#
var regex = new Regex(dollarSignReplacement: "(?:\r\n}|(?!\r\n)[\r-\n\u0085\u2028\u2029])");
```

**Comment by BurntSushi** [comment]
In the course of porting the regex engines to `regex-automata`, I've given this particular issue a lot of thought. In particular, I personally think the best path forward is to support both `\r\n` and `\n` as line endings, while ignoring the rest (`\r` on its own, as well as the rest of the line endings recommended by UTS#18). Because of compatibility, this would have to be supported via a new opt-in configuration knob. In a hypothetical regex 2.0, we could consider enabling it by default.

I've tried to think of a simple way of supporting `\r\n` in the DFA, since that is ultimately where my hesitation has been. Basically, the last thing I want to do is add another `\b`, which is currently the only thing you can put into a regex that disqualifies use of the DFA. I note that supporting these constructs in a DFA is not a _theoretical_ concern, but an engineering one. `\b` is difficult because it blows up the size of the automaton quite a bit. (Even the ASCII-only variant, which is fully supported in the DFA, has this effect.) `\r\n` doesn't really suffer from the problem of blowing up the size of the automaton much, but rather, it requires multi-byte look-around. The DFA is really only setup to handle single-byte look-around. Adding support for two-byte look-around, which would be required here, is a significant challenge. And carrying that look-around through multiple DFA states is another engineering challenge. On top of all of that, using two-byte look-around adds an additional small but complex-to-avoid cost to every search.

Overall, after giving this problem a lot of thought, I cannot see how to easily make it work in the DFA. If we accept defeat here, then supporting this means something like this:

* When CRLF-aware `$` is used, we can still _heuristically_ enable the DFA by causing it to quit and fall back to another engine if a `\r` byte is seen. (This is what we do for the Unicode word boundary currently, except any non-ASCII byte causes it to quit.)
* Support CRLF-aware `$` in the slower NFA simulation engines. "slower" in this context means an order of magnitude. You are very likely to notice this. It comes up now and then when people try to use this crate to find capture group offsets. With that said, this crate's literal optimizations can sometimes make up for this and mask performance problems in some common cases.

I'm honestly not quite sure what to do. Part of me think we should do this because correctness is important. But another part of me thinks that if this is opt-in, then it's likely folks are still going to burn time debugging problems caused by `$` _not_ being CRLF-aware by default anyway. And by the time that happens, implementing a work-around is usually not such a big deal. That is, my feeling is the key issue is that this problem manifests in the first place. On top of that, the implementation options for this aren't great unfortunately.

On the other hand, making `$` work intuitively in an dominant environment just seems like something we should obviously do, even if it silently makes things slower and even if you have to opt into it.

**Comment by BurntSushi** [comment]
I've been continuing to think about this, and I'm wondering if there is another path here. What if we _did_ treat `\r` as a new line, but _didn't_ match between a `\r` and a `\n`? I _think_ this would let us avoid needing to extra the delaying of matches from 1 byte to 2 bytes. Instead, `^` and `$` can be handled a bit like `\b` is handled: as both a look-behind and a look-ahead assertion. The key here is that if `\r` is treated a newline unto itself, then `$` is known to match at positions where `\r` follows no matter what. At that point, all we have to do is ensure that it _doesn't_ match at positions between a `\r` and `\n`. And I _think_ that's doable.

I was inspired by looking at this mailing list thread again after thinking about the problem: https://groups.google.com/g/re2-dev/c/uNi95OBEiTY

**Comment by BatmanAoD** [comment]
I think it's entirely reasonable to unconditionally consider `\r` to be the start of a line-ending, and handle the "`\n` following `\r`" situation as the actual special case. (That's basically my suggestion from back in 2017. As I mentioned then, Putty actually does use `\r` in isolation by default, so it's not an entirely obsolete form of newline.)

> it sounds like `.` needs to become `[^\r\n]`.

That sounds right to me. As a user, I can't imagine a scenario where `.` failing to match `\r` would be would be more surprising than matching would be (without a flag to have `.` match newlines, of course).

>  "Where the 'arbitrary character pattern' matches a newline sequence, it must match all of the newline sequences." Like, huh? `.` is usually specified to *not* match a newline sequence.

I assume that's exactly why they phrased it the way they did, i.e., they're talking about implementing a "multiline" mode (mentioned above the list), such as this library's `dot_matches_new_line`. The requirement (as I read it) is that in this mode, `.` matches exactly the same things that `$` would match.

**Comment by BurntSushi** [comment]
> I assume that's exactly why they phrased it the way they did, i.e., they're talking about implementing a "multiline" mode (mentioned above the list), such as this library's `dot_matches_new_line`. The requirement (as I read it) is that in this mode, `.` matches exactly the same things that `$` would match.

Ah I see. Yeah, in that case, `.` wouldn't match `\r\n`, which is what UTS#18 seems to want.

**Comment by BurntSushi** [comment]
OK, I'm happy to report that this feature should land once #656 is complete. I have it working. Would folks like to review the docs for it?

```rust
    /// Enable or disable the "CRLF mode" flag by default.
    ///
    /// By default this is disabled. It may alternatively be selectively
    /// enabled in the regular expression itself via the `R` flag.
    ///
    /// When CRLF mode is enabled, the following happens:
    ///
    /// * Unless `dot_matches_new_line` is enabled, `.` will match any character
    /// except for `\r` and `\n`.
    /// * When `multi_line` mode is enabled, `^` and `$` will treat `\r\n`,
    /// `\r` and `\n` as line terminators. And in particular, neither will
    /// match between a `\r` and a `\n`.
```

The key things to note here are:

* CRLF mode is opt-in.
* You'll be able to opt-in via `RegexBuilder::crlf`, or by using the new `R` inline flag.
* `\r` on its own is treated as a line terminator, as suggested by @BatmanAoD above. The key trick is that `^` and `$` won't match between a `\r` and a `\n`. (I would ideally rather not treat `\r` as a line terminator unto itself, but it's just not feasible to do.)
* In CRLF mode, `.` becomes `[^\r\n]` instead of `[^\n]`.

For a deeper dive, here's a smattering of tests showing CRLF mode semantics:

```toml
# This is a basic test that checks ^ and $ treat \r\n as a single line
# terminator. If ^ and $ only treated \n as a line terminator, then this would
# only match 'xyz' at the end of the haystack.
[[test]]
name = "basic"
regex = '(?mR)^[a-z]+$'
haystack = "abc\r\ndef\r\nxyz"
matches = [[0, 3], [5, 8], [10, 13]]

# Tests that a CRLF-aware '^$' assertion does not match between CR and LF.
[[test]]
name = "start-end-non-empty"
regex = '(?mR)^$'
haystack = "abc\r\ndef\r\nxyz"
matches = []

# Tests that a CRLF-aware '^$' assertion matches the empty string, just like
# a non-CRLF-aware '^$' assertion.
[[test]]
name = "start-end-empty"
regex = '(?mR)^$'
haystack = ""
matches = [[0, 0]]

# Tests that a CRLF-aware '^$' assertion matches the empty string preceding
# and following a line terminator.
[[test]]
name = "start-end-before-after"
regex = '(?mR)^$'
haystack = "\r\n"
matches = [[0, 0], [2, 2]]

# Tests that a CRLF-aware '^' assertion does not split a line terminator.
[[test]]
name = "start-no-split"
regex = '(?mR)^'
haystack = "abc\r\ndef\r\nxyz"
matches = [[0, 0], [5, 5], [10, 10]]

# Same as above, but with adjacent runs of line terminators.
[[test]]
name = "start-no-split-adjacent"
regex = '(?mR)^'
haystack = "\r\n\r\n\r\n"
matches = [[0, 0], [2, 2], [4, 4], [6, 6]]

# Same as above, but with adjacent runs of just carriage returns.
[[test]]
name = "start-no-split-adjacent-cr"
regex = '(?mR)^'
haystack = "\r\r\r"
matches = [[0, 0], [1, 1], [2, 2], [3, 3]]

# Same as above, but with adjacent runs of just line feeds.
[[test]]
name = "start-no-split-adjacent-lf"
regex = '(?mR)^'
haystack = "\n\n\n"
matches = [[0, 0], [1, 1], [2, 2], [3, 3]]

# Tests that a CRLF-aware '$' assertion does not split a line terminator.
[[test]]
name = "end-no-split"
regex = '(?mR)$'
haystack = "abc\r\ndef\r\nxyz"
matches = [[3, 3], [8, 8], [13, 13]]

# Same as above, but with adjacent runs of line terminators.
[[test]]
name = "end-no-split-adjacent"
regex = '(?mR)$'
haystack = "\r\n\r\n\r\n"
matches = [[0, 0], [2, 2], [4, 4], [6, 6]]

# Same as above, but with adjacent runs of just carriage returns.
[[test]]
name = "end-no-split-adjacent-cr"
regex = '(?mR)$'
haystack = "\r\r\r"
matches = [[0, 0], [1, 1], [2, 2], [3, 3]]

# Same as above, but with adjacent runs of just line feeds.
[[test]]
name = "end-no-split-adjacent-lf"
regex = '(?mR)$'
haystack = "\n\n\n"
matches = [[0, 0], [1, 1], [2, 2], [3, 3]]

# Tests that '.' does not match either \r or \n when CRLF mode is enabled. Note
# that this doesn't require multi-line mode to be enabled.
[[test]]
name = "dot-no-crlf"
regex = '(?R).'
haystack = "\r\n\r\n\r\n"
matches = []
```

It was _quite_ gnarly to add, and in so doing, I actually uncovered a bug in the lazy DFA (present in both the status quo and my rewrite):

```toml
# A variant of the problem described here:
# https://github.com/google/re2/blob/89567f5de5b23bb5ad0c26cbafc10bdc7389d1fa/re2/dfa.cc#L658-L667
[[test]]
name = "alt-with-assertion-repetition"
regex = '(?:\b|%)+'
haystack = "z%"
bounds = [1, 2]
anchored = true
# This was reporting [1, 2], which is
# not the correct leftmost-first match.
matches = [[1, 1]]
```
